server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java

```diff
@@ -176,7 +176,24 @@ public boolean isAggregatable() {
      */
     // TODO: Standardize exception types
     public abstract Query termQuery(Object value, @Nullable QueryShardContext context);
-
+    
+    
+    // Case insensitive form of term query (not supported by all fields so must be overridden to enable)
+    public Query termQueryCaseInsensitive(Object value, @Nullable QueryShardContext context) {
+        throw new QueryShardException(context, "[" + name + "] field which is of type [" + typeName() + 
+            "], does not support case insensitive term queries");
+    }    
+
+    
+    // Case insensitive form of terms query
+    public Query termsQueryCaseInsensitive(List<?> values, @Nullable QueryShardContext context) {        
```

Comment:
I wonder what we should do here ? We'd need to add the support for case insensitive matching in `TermInSetQuery` if we want to avoid building giant boolean query in the `keyword` field. For now I think it's ok to not provide this option on `terms` query since we cannot support them efficiently ?

Reply:
@jimczi `terms` query is currently used by eql for more obvious queries like `file where file_name in ("wininit.exe", "lsass.exe")` and (less obvious) for [`cidrMatch` function](https://www.elastic.co/guide/en/elasticsearch/reference/7.9/eql-function-ref.html#eql-fn-cidrmatch) for matching an IP address in a list of CIDR blocks. Would be useful to have case insensitive support for it as well.

Reply:
> (less obvious) for [`cidrMatch` function](https://www.elastic.co/guide/en/elasticsearch/reference/7.9/eql-function-ref.html#eql-fn-cidrmatch) for matching an IP address in a list of CIDR blocks.

Neat! I just want to make sure you aren't expanding the cidr blocks or anything - we natively support the cidr match in the `term` and `terms` query.

Reply:
Nice.

Hopefully you'll be able to use runtime field before too long!

Reply:
>For now I think it's ok to not provide this option on terms query since we cannot support them efficiently ?

If we don't support it won't people just be forced to do what we do internally and create a bool `should` array of `term` queries?
Is there a query-complexity circuit-breaker we're somehow missing here?

Reply:
Adding a +1 to have this supported in `terms` query. The alternative in EQL, in case we decide to not support this, would probably be (as @markharwood mentioned) to create a `bool` query with a bunch of `term` queries in it. CC @jimczi

Reply:
My worry is that we'd internally create an array of should `term` queries because the `TermsInSet` query doesn't handle case insensitive queries. It's maybe not a big issue but that wouldn't be consistent since a case insensitive terms query would fail with  more than 1024 terms (the max boolean clause limit) while the normal one would use the optimized `TermsInSet` query. So my take on  this is that we should implement a proper support in `TermInSet` if we really want to handle `terms` case insensitive query. For EQL and SQL I don't think there's a real need to handle `terms` though. The `in` operator doesn't need to handle thousands of terms so imo that would be acceptable to translate it into an array of `term` query. Same for the `cidrMatch` function that also work on `term` queries.

Reply:
OK - I'll pull the ` termsQueryCaseInsensitive` method for now. This PR probably has enough changes to consider already